lib/segment/src/index/field_index/immutable_point_to_values.rs

```diff
@@ -0,0 +1,101 @@
+use std::ops::Range;
+
+use common::types::PointOffsetType;
+
+// Flatten points-to-values map
+// It's an analogue of `Vec<Vec<N>>` but more RAM efficient because it stores values in a single Vec.
+// This structure doesn't support adding new values, only removing.
+// It's used in immutable field indices like `ImmutableMapIndex`, `ImmutableNumericIndex`, etc to store points-to-values map.
+#[derive(Debug, Clone, Default)]
+pub struct ImmutablePointToValues<N: Default> {
+    // ranges in `point_to_values_container` which contains values for each point
+    // `u32` is used instead of `usize` because it's more RAM efficient
+    // We can expect that we will never have more than 4 billion values per segment
+    point_to_values: Vec<Range<u32>>,
+    // flattened values
+    point_to_values_container: Vec<N>,
+}
+
+impl<N: Default> ImmutablePointToValues<N> {
+    pub fn new(src: Vec<Vec<N>>) -> Self {
+        let mut point_to_values: Vec<_> = Default::default();
+        let mut point_to_values_container: Vec<_> = Default::default();
+        for values in src {
+            let values = values.into_iter().collect::<Vec<_>>();
+            let container_len = point_to_values_container.len() as u32;
+            let range = container_len..container_len + values.len() as u32;
+            point_to_values.push(range.clone());
+            point_to_values_container.extend(values);
```

Comment:
```suggestion
            let container_len = point_to_values_container.len() as u32;
            let range = container_len..container_len + values.len() as u32;
            point_to_values.push(range.clone());
            point_to_values_container.extend(values);
```

Haven't tested this, but this would prevent an intermediate `Vec`. [`extend`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend) can take an `IntoIterator`.

Reply:
You are right! Thanks for this finding, this intermediate `Vec` is not needed. Fixed